plugins/inputs/clickhouse/clickhouse.go

```diff
@@ -0,0 +1,355 @@
+			if err := ch.commonMetrics(acc, &conn, metric); err != nil {
+				acc.AddError(err)
+			}
+		}
+	}
+	return nil
+}
+
+func (ch *ClickHouse) clusterIncludeExcludeFilter() string {
+	if len(ch.ClusterInclude) == 0 && len(ch.ClusterExclude) == 0 {
+		return ""
+	}
+	var (
+		escape = func(in string) string {
+			return "'" + strings.NewReplacer(`\`, `\\`, `'`, `\'`).Replace(in) + "'"
+		}
+		makeFilter = func(expr string, args []string) string {
+			in := make([]string, 0, len(args))
+			for _, v := range args {
+				in = append(in, escape(v))
+			}
+			return "cluster " + expr + " (" + strings.Join(in, ", ") + ")"
+		}
+	)
+
+	if len(ch.ClusterInclude) != 0 {
+		return "WHERE " + makeFilter("IN", ch.ClusterInclude)
+	}
+
+	return "WHERE " + makeFilter("NOT IN", ch.ClusterExclude)
+}
+
+func (ch *ClickHouse) commonMetrics(acc telegraf.Accumulator, conn *connect, metric string) error {
+	var result []struct {
+		Metric string   `json:"metric"`
+		Value  chUInt64 `json:"value"`
+	}
+	if err := ch.execQuery(conn.url, commonMetrics[metric], &result); err != nil {
+		return err
+	}
+
+	tags := map[string]string{
+		"source": conn.Hostname,
+	}
+	if len(conn.Cluster) != 0 {
+		tags["cluster"] = conn.Cluster
+	}
+	if conn.ShardNum != 0 {
+		tags["shard_num"] = strconv.Itoa(conn.ShardNum)
+	}
+
+	fields := make(map[string]interface{})
+	for _, r := range result {
+		fields[internal.SnakeCase(r.Metric)] = uint64(r.Value)
+	}
+
+	acc.AddFields("clickhouse_"+metric, fields, tags)
+
+	return nil
+}
+
+func (ch *ClickHouse) tables(acc telegraf.Accumulator, conn *connect) error {
+	var parts []struct {
+		Database string   `json:"database"`
+		Table    string   `json:"table"`
+		Bytes    chUInt64 `json:"bytes"`
+		Parts    chUInt64 `json:"parts"`
+		Rows     chUInt64 `json:"rows"`
+	}
+
+	if err := ch.execQuery(conn.url, systemParts, &parts); err != nil {
+		return err
+	}
+	tags := map[string]string{
+		"source": conn.Hostname,
+	}
+	if len(conn.Cluster) != 0 {
+		tags["cluster"] = conn.Cluster
+	}
+	if conn.ShardNum != 0 {
+		tags["shard_num"] = strconv.Itoa(conn.ShardNum)
+	}
+	for _, part := range parts {
+		tags["table"] = part.Table
+		tags["database"] = part.Database
+		acc.AddFields("clickhouse_tables",
+			map[string]interface{}{
+				"bytes": uint64(part.Bytes),
+				"parts": uint64(part.Parts),
+				"rows":  uint64(part.Rows),
+			},
+			tags,
+		)
+	}
+	return nil
+}
+
+type clickhouseError struct {
+	StatusCode int
+	body       []byte
+}
+
+func (e *clickhouseError) Error() string {
+	return fmt.Sprintf("ClickHouse server returned an error code: %d\n%s", e.StatusCode, e.body)
+}
+
+func (ch *ClickHouse) execQuery(url *url.URL, query string, i interface{}) error {
+	q := url.Query()
+	q.Set("query", query+" FORMAT JSON")
+	url.RawQuery = q.Encode()
+	resp, err := ch.client.Get(url.String())
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode >= 300 {
+		body, _ := ioutil.ReadAll(resp.Body)
```

Comment:
can use a LimitedReader to read at most 200 bytes here.

Reply:
won't fix, this fix is necessary for complete review?